### ISSUE TITLE
[ENH] Make freelists atomic on the freethreading build

### ISSUE THREAD

**Issue by da-woods** [cython/cython#6433]
### Describe your issue

Freelists are currently turned off on the freethreading build because they aren't thread safe. The could be re-enabled if the count was atomic.

However, it'd also be worth checking if they're still a worthwhile optimization in that case.

**Comment by scoder** [comment]
> worth checking if they're still a worthwhile optimization in that case

Yeah, that's not certain. If they end up in memory buckets of other threads, then we'd potentially lose a lot of performance.

I'd rather keep them disabled all together, until there's a reason to believe that they are worthwhile. Free-threaded Python lives from the assumption that a bit of per-core performance is worth trading for better parallel performance. Efficient parallel freelists probably need a more sophisticated implementation all together.

**Comment by da-woods** [comment]
There might be an argument for making them threadsafe, but leaving `CYTHON_USE_FREELISTS` off by default. That at least makes it for people to turn on if they want.
